src/contract.rs

```diff
@@ -914,9 +1014,163 @@ fn query_validators_metrics_by_timestamp(
     Ok(res)
 }
 
+// off chain code
+
+//todo: test, validate
+fn remove_off_chain_metrics_for_timestamp(
+    deps: DepsMut,
+    info: MessageInfo,
+    timestamp: u64,
+) -> Result<Response, ContractError> {
+    let mut timestamp_existed = true;
+
+    if !sender_is_manager(&deps, &info) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if !OFF_CHAIN_TIMESTAMPS.has(deps.storage, U64Key::from(timestamp)) {
+        timestamp_existed = false;
+    }
+
+    OFF_CHAIN_TIMESTAMPS.remove(deps.storage, U64Key::from(timestamp));
+    OFF_CHAIN_TIMESTAMP_META_DATA.remove(deps.storage, U64Key::from(timestamp));
+
+    let validator_idxs_to_remove: Vec<u16> = OFF_CHAIN_STATE_FOR_VALIDATOR
+        .prefix(U64Key::from(timestamp))
+        .range(deps.storage, Option::None, Option::None, Order::Ascending)
+        .map(|item| return item.unwrap().1.validator_idx)
+        .collect();
+
+    validator_idxs_to_remove.into_iter().for_each(|idx| {
+        OFF_CHAIN_STATE_FOR_VALIDATOR
+            .remove(deps.storage, (U64Key::from(timestamp), U16Key::from(idx)))
+    });
+
+    Ok(Response::new()
+        .add_attribute("method", "remove_off_chain_metrics_for_timestamp")
+        .add_attribute(
+            "message",
+            if timestamp_existed {
+                "timestamp successfully removed"
```

Comment:
Understanding when this tx will be done is also useful especially as we paginate.

Reply:
Yeah, will provide functionality to paginate this (if required). Currently just providing this as a msg if we want to clean up certain timestamps. Let's see how costly it is actually.